fuels-core/src/json_abi.rs

```diff
@@ -516,41 +516,52 @@ pub fn parse_param(param: &Property) -> Result<ParamType, Error> {
         // Simple case (primitive types, no arrays, including string)
         Ok(param_type) => Ok(param_type),
         Err(_) => {
-            match param.type_field.contains("struct") || param.type_field.contains("enum") {
-                true => Ok(parse_custom_type_param(param)?),
-                false => {
-                    match param.type_field.contains('[') && param.type_field.contains(']') {
-                        // Try to parse array (T[M]) or string (str[M])
-                        true => Ok(parse_array_param(param)?),
-                        // Try to parse enum or struct
-                        false => Ok(parse_custom_type_param(param)?),
-                    }
+            if param.type_field.contains("struct") || param.type_field.contains("enum") {
+                return parse_custom_type_param(param);
+            }
+            if param.type_field.contains('[') && param.type_field.contains(']') {
+                // Try to parse array ([T; M]) or string (str[M])
+                if param.type_field.contains("str[") {
+                    return parse_string_param(param);
                 }
+                return parse_array_param(param);
             }
+            // Try to parse enum or struct
+            return parse_custom_type_param(param);
         }
     }
 }
 
-pub fn parse_array_param(param: &Property) -> Result<ParamType, Error> {
-    // Split "T[n]" string into "T" and "[n]"
+pub fn parse_string_param(param: &Property) -> Result<ParamType, Error> {
+    // Split "str[n]" string into "str" and "[n]"
     let split: Vec<&str> = param.type_field.split('[').collect();
-    if split.len() != 2 {
-        return Err(Error::MissingData(format!(
-            "invalid parameter type: {}",
+    if split.len() != 2 || !split[0].eq("str") {
+        return Err(Error::InvalidType(format!(
+            "Expected parameter type `str[n]`, found `{}`",
             param.type_field
         )));
     }
-
-    let param_type = ParamType::from_str(split[0]).unwrap();
-
     // Grab size in between brackets, i.e the `n` in "[n]"
     let size: usize = split[1][..split[1].len() - 1].parse().unwrap();
+    Ok(ParamType::String(size))
+}
 
-    if let ParamType::String(_) = param_type {
-        Ok(ParamType::String(size))
-    } else {
-        Ok(ParamType::Array(Box::new(param_type), size))
+pub fn parse_array_param(param: &Property) -> Result<ParamType, Error> {
+    // Split "[T; n]" string into "T" and "n"
+    let split: Vec<&str> = param.type_field.split("; ").collect();
+    if split.len() != 2 {
+        return Err(Error::InvalidType(format!(
+            "Expected parameter type `[T; n]`, found `{}`",
+            param.type_field
+        )));
     }
+    let (type_field, size) = (split[0], split[1]);
+    let type_field = type_field[1..].to_string();
+    let param_type = ParamType::from_str(&type_field).unwrap();
```

Comment:
We should avoid using `unwrap()` when it's possible / not too hard to avoid it. In this case here, it's not too hard to change `unwrap()` to `?` to better handle the error and avoid panicking. The error from parsing a bad `type_field` would be very cryptic to the end-user that could've misspelled the type in the JSON ABI.

Reply:
I'm still not used to this you're absolutely right